source/robot_model/src/Model.cpp

```diff
@@ -29,17 +47,145 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+      std::string path = matches_file[1];
+      
+      if (std::regex_search(path, matches_package_name, package_name_pattern)) {
+          package_names.insert(matches_package_name[1]);
+      }
+      
+      start = matches_file[0].second;   
+  }
+
+  std::vector<std::string> package_paths;
+  for (const auto& package_name : package_names) {
+    if (meshloader_callback_){
+      auto package_path = (*this->meshloader_callback_)(package_name);
+      auto target = "package://" + package_name + "/";
+      auto replacement = package_path;
+      size_t start_position = 0;
+      while ((start_position = urdf.find(target, start_position)) != std::string::npos) {
+          // Replace the target with the replacement string
+          urdf.replace(start_position, target.length(), replacement);
+          // Move past the last replacement
+          start_position += replacement.length();
+      }
+      package_paths.push_back(package_path);
+    }
+  }
+  return package_paths;
+}
+
+void Model::init_model(bool load_collision_geometries) {
+  std::ifstream file_stream(this->get_urdf_path());
+  if (!file_stream.is_open()) {
+      throw std::runtime_error("Unable to open file: " + this->get_urdf_path());
+  }
+  std::stringstream buffer;
+  buffer << file_stream.rdbuf();
+  auto urdf = buffer.str();
+
+  pinocchio::urdf::buildModelFromXML(urdf, this->robot_model_);
   this->robot_data_ = pinocchio::Data(this->robot_model_);
+
+  if (load_collision_geometries) {
+    this->init_geom_model(urdf);
+  }
+
   // get the frames
   std::vector<std::string> frames;
   for (auto& f : this->robot_model_.frames) {
     frames.push_back(f.name);
   }
   // remove universe and root_joint frame added by Pinocchio
   this->frames_ = std::vector<std::string>(frames.begin() + 2, frames.end());
-  this->init_qp_solver();
+  this->init_qp_solver();  
+}
+
+// Method to initialize collision geometries
+void Model::init_geom_model(std::string urdf) {
+    try {
+      auto package_paths = this->resolve_package_paths_in_urdf(urdf);
+      pinocchio::urdf::buildGeom(this->robot_model_,
+                                std::istringstream(urdf),
+                                pinocchio::COLLISION, 
+                                this->geom_model_, 
+                                package_paths);
+      this->geom_model_.addAllCollisionPairs();
+    
+      std::vector<pinocchio::CollisionPair> excluded_pairs = this->generate_joint_exclusion_list();
+
+      // remove collision pairs for linked joints (i.e. parent-child joints)
+      for (const auto& pair : excluded_pairs) {
+          this->geom_model_.removeCollisionPair(pair);
+      }
+
+      this->geom_data_ = pinocchio::GeometryData(this->geom_model_);
+    } catch (const std::exception& e) {
+        throw robot_model::exceptions::CollisionGeometryException("Failed to initialize Geomerty model for " + this->get_robot_name());
+    }
```

Comment:
You have a typo here. Would also be nice to forward the base exception message.
```suggestion
    } catch (const std::exception& ex) {
        throw robot_model::exceptions::CollisionGeometryException("Failed to initialize geometry model for " + this->get_robot_name() + ": " + ex.what());
    }
```